src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java

```diff
@@ -193,39 +207,64 @@ public void onEnterWaiting(Queue.WaitingItem wi) {
             }
             final long taskId = wi.getId();
             final Job<?,?> job = (Job) wi.task;
-            // TODO would actually be better to use GitHubSCMSource.retrieve(SCMHead, TaskListener) (when properly implemented).
-            // That would allow us to find the current head commit even for non-PR jobs.
+            final SCMSource source = SCMSource.SourceByItem.findSource(job);
+            if (!(source instanceof GitHubSCMSource)) {
+                return;
+            }
             final SCMHead head = SCMHead.HeadByItem.findHead(job);
             if (!(head instanceof PullRequestSCMHead)) {
                 return;
             }
+            // prevent delays in the queue when updating github
             Computer.threadPoolForRemoting.submit(new Runnable() {
                 @Override
                 public void run() {
+                    GitHub gitHub = null;
                     try {
-                        GHRepository repo = lookUpRepo(job);
-                        if (repo != null) {
-                            int number = ((PullRequestSCMHead) head).getNumber();
-                            GHPullRequest pr = repo.getPullRequest(number);
-                            String url = DisplayURLProvider.get().getJobURL(job);
-                            // Has not been built yet, so we can only guess that the current PR head is what will be built.
-                            // In fact the submitter might push another commit before this build even starts.
-                            if (Jenkins.getActiveInstance().getQueue().getItem(taskId) instanceof Queue.LeftItem) {
-                                // we took too long and the item has left the queue, no longer valid to apply pending
-                                // status. JobCheckOutListener is now responsible for setting the pending status.
+                        gitHub = lookUpGitHub(job);
+                        try {
+                            if (gitHub == null || gitHub.rateLimit().remaining < 8) {
+                                // we are an optimization to signal commit status early, no point waiting for
+                                // the rate limit to refresh as the checkout will ensure the status is set
+                                return;
+                            }
+                            String hash = resolveHeadCommit(source.fetch(head, null));
```

Comment:
Not really related to the main point of this PR, is it? Seems you are sneaking in a separate RFE. (Useful but confusing here.)